sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala

```diff
@@ -0,0 +1,195 @@
+          // whenever the current fills up.
+          // Each row is laid out in memory as [rowSize (int)|rowData (rowSize bytes)].
+          // The end of the block is marked by a special rowSize, END_OF_BLOCK (-1).
+          val block = taskMemoryManager.allocateUnchecked(blockSize)
+
+          var currOffset = 0
+          while (bufferedRowIterator.hasNext &&
+            currOffset + 4 + bufferedRowIterator.head.getSizeInBytes < blockSize - 4) {
+            val currRow = bufferedRowIterator.head
+            Platform.putInt(
+              block.getBaseObject, block.getBaseOffset + currOffset, currRow.getSizeInBytes)
+            currRow.writeToMemory(block.getBaseObject, block.getBaseOffset + currOffset + 4)
+            bufferedRowIterator.next()
+            currOffset += 4 + currRow.getSizeInBytes
+          }
+          Platform.putInt(block.getBaseObject, block.getBaseOffset + currOffset, END_OF_BLOCK)
+
+          compressionCodec match {
+            case Some(codec) => compressBlock(block, codec, TaskContext.get().taskMemoryManager())
+            case None => block
+          }
+        }
+
+        def hasNext: Boolean = bufferedRowIterator.hasNext
+      }
+    }
+    .setName(compressionType.getOrElse("") + "_" + child.nodeName)
+    .persist(StorageLevel.MEMORY_ONLY)
+    // TODO(josh): is this the right name?
+
+    cachedRDD.mapPartitions { blockIterator =>
+      val compressionCodec: Option[CompressionCodec] =
+        compressionType.map { t => CompressionCodec.createCodec(SparkEnv.get.conf, t) }
+      blockIterator.flatMap { rawBlock =>
+        val block: MemoryBlock = compressionCodec match {
+          case Some(codec) => decompressBlock(rawBlock, blockSize, codec)
+          case None => rawBlock
+        }
+
+        new Iterator[UnsafeRow] {
+          private[this] val unsafeRow = new UnsafeRow()
+          private[this] var currOffset: Long = 0
+          private[this] var _nextRowSize = getNextRowSize()
+          private[this] def getNextRowSize(): Int = {
+            if (currOffset >= block.size()) {
+              -1
+            } else {
+              Platform.getInt(block.getBaseObject, block.getBaseOffset + currOffset)
+            }
+          }
+          override def hasNext: Boolean = _nextRowSize != END_OF_BLOCK
+          override def next(): UnsafeRow = {
+            assert(_nextRowSize > 0)
+            currOffset += 4
+            unsafeRow.pointTo(
+              block.getBaseObject, block.getBaseOffset + currOffset, numFields, _nextRowSize)
+            currOffset += _nextRowSize
+            _nextRowSize = getNextRowSize()
+            unsafeRow
+          }
+        }
+      }
+    }
+  }
+
+  private def compressBlock(
+      memoryBlock: MemoryBlock,
+      compressionCodec: CompressionCodec,
+      taskMemoryManager: TaskMemoryManager): MemoryBlock = {
+    // Compress the block using an on-heap byte array
+    val compressedBlockArray: Array[Byte] = {
```

Comment:
Can we also reuse this `Array` across `MemoryBlock`s since L162 copies it off-heap?